File: DSL/com.larsreimann.safeds/src/testFixtures/kotlin/de/unibonn/simpleml/testing/TestRanges.kt
package com.larsreimann.safeds.testing

import de.unibonn.simpleml.location.XtextPosition
import de.unibonn.simpleml.location.XtextRange
import de.unibonn.simpleml.testing.FindTestRangesResult.CloseWithoutOpenFailure
import de.unibonn.simpleml.testing.FindTestRangesResult.OpenWithoutCloseFailure
import de.unibonn.simpleml.testing.FindTestRangesResult.Success
import de.unibonn.simpleml.testing.TestMarker.CLOSE
import de.unibonn.simpleml.testing.TestMarker.OPEN

/**
 * Finds test ranges, i.e. parts of the program delimited by opening and closing test markers. They are sorted by the
 * position of their opening test markers. In case opening and closing markers don't match an error value is returned.
 * Nested test markers are supported.
 *
 * @param program The program with test markers.
 * @return A wrapper that indicates success of failure.
 * @see FindTestRangesResult
 * @see TestMarker
 */
fun findTestRanges(program: String): FindTestRangesResult {
    var currentLine = 1
    var currentColumn = 1
    var previousChar: Char? = null

    val testRangeStarts = ArrayDeque<TestRangeStart>()
    val finishedLocations = mutableListOf<XtextRange>()

    program.toCharArray().forEachIndexed { currentIndex, currentChar ->
        when (currentChar) {
            OPEN -> {
                currentColumn++

                testRangeStarts.addLast(
                    TestRangeStart(currentLine, currentColumn, currentIndex)
                )
            }
            CLOSE -> {
                currentColumn++

                if (testRangeStarts.isEmpty()) {
                    return CloseWithoutOpenFailure(
                        XtextPosition.fromInts(
                            line = currentLine,
                            column = currentColumn - 1
                        )
                    )
                }

                finishedLocations += testRangeStarts.removeLast().toProgramRange(
                    endLine = currentLine,
                    endColumn = currentColumn - 1,
                    endIndex = currentIndex - 1
                )
            }
            '\r' -> {
                currentLine++
                currentColumn = 1
            }
            '\n' -> {
                if (previousChar != '\r') {
                    currentLine++
                    currentColumn = 1
                }
            }
            else -> {
                currentColumn++
            }
        }

        previousChar = currentChar
    }

    return when {
        testRangeStarts.isEmpty() -> Success(finishedLocations.sortedBy { it.start })
        else -> OpenWithoutCloseFailure(
            testRangeStarts.map {
                XtextPosition.fromInts(it.startLine, it.startColumn - 1)
            }
        )
    }
}

/**
 * A wrapper that indicates success of failure of the `findTestRanges` method.
 */
@Suppress("MemberVisibilityCanBePrivate")
sealed interface FindTestRangesResult {

    /**
     * Opening and closing test markers matched and program ranges were successfully created.
     */
    class Success(val ranges: List<XtextRange>) : FindTestRangesResult

    /**
     * Something went wrong when creating program ranges.
     */
    sealed interface Failure : FindTestRangesResult {

        /**
         * A human-readable description of what went wrong.
         */
        val message: String
    }

    /**
     * Found a closing test marker without a previous opening test marker.
     */
    class CloseWithoutOpenFailure(val position: XtextPosition) : Failure {
        override val message: String
            get() {
                return "Found '$CLOSE' without previous '$OPEN' at $position."
            }
    }

    /**
     * Reached the end of the program but there were still unclosed opening test markers.
     */
    class OpenWithoutCloseFailure(val positions: List<XtextPosition>) : Failure {
        override val message: String
            get() {
                return "Found '$OPEN' without following '$CLOSE' at ${positions.joinToString()}."
            }
    }
}

/**
 * Stores where a test range starts.
 */
private class TestRangeStart(val startLine: Int, val startColumn: Int, val startIndex: Int) {
    fun toProgramRange(endLine: Int, endColumn: Int, endIndex: Int): XtextRange {
        return XtextRange.fromInts(
            startLine,
            startColumn,
            endLine,
            endColumn,
            length = endIndex - startIndex
        )
    }
}

File: DSL/tests/helpers/testRanges.ts
import { Result } from 'true-myth';
import { Range, Position } from 'vscode-languageserver';
import { CLOSE, OPEN } from './testMarker';
import {positionToString} from "./stringification";

/**
 * Finds test ranges, i.e. parts of the program delimited by opening and closing test markers. They are sorted by the
 * position of their opening test markers. In case opening and closing markers don't match, an error value is returned.
 * Nested test markers are supported.
 *
 * @param program The program with test markers.
 * @return A wrapper that indicates success of failure.
 * @see FindTestRangesError
 * @see CLOSE
 * @see OPEN
 */
export const findTestRanges = (program: string): Result<Range[], FindTestRangesError> => {
    let currentLine = 0;
    let currentColumn = 0;
    let previousChar: string | null = null;

    const testRangeStarts: Position[] = [];
    const finishedLocations: Range[] = [];

    for (let currentIndex = 0; currentIndex < program.length; currentIndex++) {
        const currentChar = program[currentIndex];
        switch (currentChar) {
            case OPEN:
                currentColumn++;

                testRangeStarts.push(Position.create(currentLine, currentColumn));
                break;
            case CLOSE:
                currentColumn++;

                if (testRangeStarts.length === 0) {
                    return Result.err(new CloseWithoutOpenError(Position.create(currentLine, currentColumn - 1)));
                }

                finishedLocations.push(
                    Range.create(testRangeStarts.pop()!, Position.create(currentLine, currentColumn - 1)),
                );
                break;
            case '\r':
                currentLine++;
                currentColumn = 0;
                break;
            case '\n':
                if (previousChar !== '\r') {
                    currentLine++;
                    currentColumn = 0;
                }
                break;
            default:
                currentColumn++;
        }

        previousChar = currentChar;
    }

    if (testRangeStarts.length > 0) {
        return Result.err(
            new OpenWithoutCloseError(
                testRangeStarts.map((position) => Position.create(position.line, position.character - 1)),
            ),
        );
    } else {
        return Result.ok(
            finishedLocations.sort((a, b) => {
                return a.start.line === b.start.line
                    ? a.start.character - b.start.character
                    : a.start.line - b.start.line;
            }),
        );
    }
};

/**
 * Something went wrong when creating program ranges.
 */
export type FindTestRangesError = CloseWithoutOpenError | OpenWithoutCloseError;

/**
 * Found a closing test marker without a previous opening test marker.
 */
export class CloseWithoutOpenError extends Error {
    constructor(readonly position: Position) {
        super(`Found '${CLOSE}' without previous '${OPEN}' at ${position.line}:${position.character}.`);
    }
}

/**
 * Reached the end of the program but there were still unclosed opening test markers.
 */
export class OpenWithoutCloseError extends Error {
    constructor(readonly positions: Position[]) {
        super(
            `Found '${OPEN}' without following '${CLOSE}' at ${positions
                .map(positionToString)
                .join(', ')}.`,
        );
    }
}
